modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,549 @@
+                    }
+                });
+            }
+        });
+
+        return locMetasChkFut;
+    }
+
+    /** Phase 1 end. */
+    private void reducePreparationAndMetasCheck(
+        UUID procId,
+        Map<UUID, CheckResultDTO> results,
+        Map<UUID, Throwable> errors
+    ) {
+        SnapshotCheckProcessRequest locReq = currentRequest(snpName(results), procId);
+
+        assert locReq == null || locReq.opCoordId != null;
+
+        if (locReq == null || stopAndCleanOnError(locReq, errors) || !kctx.localNodeId().equals(locReq.opCoordId))
+            return;
+
+        Throwable stopClusterProcErr = null;
+
+        try {
+            assert !F.isEmpty(results) || !F.isEmpty(errors) || locReq.error() != null;
+
+            if (locReq.error() != null)
+                throw locReq.error();
+
+            locReq.metas = new HashMap<>();
+
+            Map<ClusterNode, Exception> errs = new HashMap<>();
+
+            results.forEach((nodeId, nodeRes) -> {
+                // A node might be non-baseline (not required).
+                if (locReq.nodes.contains(nodeId)) {
+                    assert nodeRes != null && nodeRes.partsHashes == null;
+                    assert kctx.cluster().get().node(nodeId) != null;
+
+                    locReq.metas.put(kctx.cluster().get().node(nodeId), nodeRes.metas);
+                }
+            });
+
+            errors.forEach((nodeId, nodeErr) -> {
+                assert locReq.nodes.contains(nodeId);
+                assert kctx.cluster().get().node(nodeId) != null;
+                assert nodeErr != null;
+
+                errs.put(kctx.cluster().get().node(nodeId), asException(nodeErr));
+            });
+
+            SnapshotMetadataVerificationTaskResult metasRes = new SnapshotMetadataVerificationTaskResult(
+                locReq.metas,
+                SnapshotChecker.reduceMetasResults(locReq.snapshotName(), locReq.snapshotPath(), locReq.metas, errs,
+                    kctx.cluster().get().localNode().consistentId())
+            );
+
+            if (!F.isEmpty(metasRes.exceptions()))
+                stopClusterProcErr = new IgniteSnapshotVerifyException(metasRes.exceptions());
+        }
+        catch (Throwable th) {
+            stopClusterProcErr = th;
+        }
+
+        if (stopClusterProcErr != null)
+            locReq.error(stopClusterProcErr);
+
+        phase2PartsHashes.start(procId, locReq);
+
+        if (log.isDebugEnabled())
+            log.debug("Started partitions validation as part of the snapshot checking. Request=" + locReq + '.');
+    }
+
+    /** Finds current snapshot name from the metas. */
+    private @Nullable String snpName(@Nullable Map<UUID, CheckResultDTO> results) {
+        if (F.isEmpty(results))
+            return null;
+
+        // Ensure the same snapshot name or empty.
+        assert F.flatCollections(results.values().stream().filter(r -> r != null && r.metas != null).map(res -> res.metas)
```

Comment:
It looks complicated. I suppose it can be check during map phase. Or, maybe, while constructing `CheckResultDTO`?